REPL prints long results

som-snytt opened the pull request:
Previously, for long results, REPL
would print `a + b + etc`. It seems
`s"${a}${b}etc` is also a tree too large
for refchecks conversion; so just
use a string builder to build the
string to print.

Fixes scala/bug#10956

som-snytt committed:
REPL prints long results

Previously, when REPL has many results
to output, it would print `a + b + etc`,
a potentially large expression tree.
`s"${a}${b}etc` is also a tree too large
for refchecks conversion; so just
use a string builder to build the
string to print, if there is more
than one result.

hrhino commented on src/repl/scala/tools/nsc/interpreter/IMain.scala:
```diff
@@ -856,12 +856,13 @@ class IMain(val settings: Settings, parentClassLoaderOverride: Option[ClassLoade
+             |    val sb = new _root_.scala.StringBuilder; """.stripMargin) // the result extraction code will emit code to append strings to this string builder
 
-          contributors map (_.resultExtractionCode(this)) foreach code.println
+             contributors.map(x => s"""sb.append("" ${x.resultExtractionCode(this)});""") foreach code.println
```
is there a reason to generate `"" + rex` rather than just calling `sb.append(rex)`?

som-snytt replied:
Too lazy to fix the weirdness.

hrhino reviewed (approved):
Very elegant, to the extent that that word means anything in the `repl` project.

Thanks!

hrhino commented:
I notice that this is a 2.13 PR. Is there a reason it can't go into 2.12? Asking only because it harshes sbt's vibe with `consoleProject` on big PRs, and IIRC they're not going to move sbt to 2.13 any time soon. (Sorry to suggest yet another sbt dispensation.)

som-snytt commented:
@hrhino Thanks! That is high praise indeed. Is there an analogue IRL? Well, this past weekend, my spouse had reserved a campsite -- I'm a hard-core backpacker, so this could be of no interest to me, right? except it turned out to be a hike-in tent site, and I hadn't brought the backpacks. However, I did have my company swag backpack, which was oddly large and square for a Macbook. The camp chair (which we didn't use b/c mosquitoes and no bug repellent) fit inside and the tent fit on top, with the straps hooked through the little loop used for hanging the backpack. My foam bedroll and sleeping hung from it in the usual way. I didn't think of it for the hike in, which was hot and miserable, but next morning, I did arrive at that solution on the second try. It was a cool, breezy morning under a marine layer, and I didn't die before I reached the car. So the analogue is apt.

som-snytt commented:
@hrhino I'd be glad to dispense one more for sbt if there were an sbt person who could put me up some weekend in NY metro area. I'll be passing thru to Maine before end of summer. That's to visit my mother one last time, if that's any incentive.

som-snytt commented:
Just noticed that wait a sec, interpolation was the elegant solution, manually string building is the ugly we are just manual laborers solution, not that there's anything wrong with manual labor unless your job as a computer scientist is to reduce manual labor and improve the lives of the many.

som-snytt commented:
Nuance, only use string builder with multiple results to print.

som-snytt committed:
No truncation in REPL tests

As they say when R2D2 is trying to shut down the trash compactor,
Shut them all down!

som-snytt commented on src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala:
```diff
@@ -1015,6 +1015,24 @@ class ILoop(config: ShellConfig, inOverride: BufferedReader = null,
+        val p = sys.Prop[Int]("wtf")
+        p.set("0")
+        p
+      }
```
Adriaaaaan! ... aaan ... aaan ... aan ...

som-snytt commented:
Rebased, and fixed second commit to respect settings in tests.

adriaanm reviewed (commented):
Beauteous.

som-snytt commented:
Maybe less necessary after the flatter tree initiative.
